src/main/groovy/ch/ergon/gradle/goodies/versioning/jgit/JGitDescribe.groovy

```diff
@@ -0,0 +1,125 @@
+package ch.ergon.gradle.goodies.versioning.jgit
+
+
+import org.eclipse.jgit.api.Git
+import org.eclipse.jgit.lib.Constants
+import org.eclipse.jgit.lib.ObjectId
+import org.eclipse.jgit.lib.Repository
+import org.eclipse.jgit.revwalk.RevCommit
+import org.eclipse.jgit.revwalk.RevWalk
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+/**
+ * JGit implementation of git describe with required flags. JGit support for describe is minimal and there is no support
+ * for --first-parent behavior.
+ * Heavily influenced by https://github.com/palantir/gradle-git-version.
+ */
+class JGitDescribe implements GitDescribe {
+    private static final Logger log = LoggerFactory.getLogger(JGitDescribe.class)
+    private final Git git
+
+    JGitDescribe(Git git) {
+        this.git = git
+    }
+
+    @Override
+    String describe(Map options) {
+        try {
+            ObjectId headObjectId = git.repository.resolve(Constants.HEAD)
+            String abbrevHead = git.repository.newObjectReader().abbreviate(headObjectId).name()
+
+            List<String> revs = revList(headObjectId)
+
+            Map<String, RefWithTagName> commitHashToTag = mapCommitsToTags(git)
+
+            for (int depth = 0; depth < revs.size(); depth++) {
+                String rev = revs.get(depth)
+                if (commitHashToTag.containsKey(rev)) {
+                    String exactTag = commitHashToTag[rev].tag
+                    if (exactTag.startsWith(options.prefix ?: "")) {
+                        String result = String.format("%s-%s-g%s", exactTag, depth, abbrevHead)
+                        if (!options.longFormat == true) {
+                            result = depth == 0 ? exactTag : result
+                        }
+                        def status = git.status().call()
+                        return status.isClean() ? result : String.format("%s-dirty", result)
+                    }
+                }
+            }
+
+            return abbrevHead
+
+        } catch (Exception e) {
+            log.debug("JGit describe failed with {}", e)
+            return null
+        }
+    }
+
+    // Mimics 'git rev-list --first-parent <commit>'
+    private List<String> revList(ObjectId initialObjectId) throws IOException {
```

Comment:
Could that not be done lazily?
If I understand code correctly, it always traverses the full history which may be problematic for repositories with long history?

Reply:
Hmm, that is a good point. I'll have to think about it...